code/datums/components/manual_heart.dm

```diff
@@ -0,0 +1,170 @@
+	pump_action.Grant(parent)
+
+	var/mob/living/carbon/carbon_parent = parent
+	var/obj/item/organ/internal/heart/parent_heart = carbon_parent.get_organ_slot(ORGAN_SLOT_HEART)
+	if(parent_heart && !HAS_TRAIT(carbon_parent, TRAIT_NOBLOOD) && carbon_parent.stat != DEAD)
+		START_PROCESSING(SSdcs, src)
+		last_pump = world.time
+
+	to_chat(parent, span_userdanger("Your heart no longer beats automatically! You have to pump it manually - otherwise you'll die!"))
+
+/datum/component/manual_heart/Destroy()
+	to_chat(parent, span_userdanger("You feel your heart start beating normally again!"))
+	var/mob/living/carbon/carbon_parent = parent
+	if(istype(carbon_parent))
+		carbon_parent.remove_client_colour(/datum/client_colour/manual_heart_blood)
+	QDEL_NULL(pump_action)
+	return ..()
+
+/datum/component/manual_heart/RegisterWithParent()
+	RegisterSignal(parent, COMSIG_CARBON_LOSE_ORGAN, PROC_REF(check_removed_organ))
+	RegisterSignal(parent, COMSIG_CARBON_GAIN_ORGAN, PROC_REF(check_added_organ))
+	RegisterSignals(parent, list(COMSIG_LIVING_DEATH, SIGNAL_ADDTRAIT(TRAIT_NOBLOOD)), PROC_REF(pause))
+	RegisterSignals(parent, list(COMSIG_LIVING_REVIVE, SIGNAL_REMOVETRAIT(TRAIT_NOBLOOD)), PROC_REF(restart))
+
+/datum/component/manual_heart/UnregisterFromParent()
+	UnregisterSignal(parent, list(COMSIG_CARBON_GAIN_ORGAN, COMSIG_CARBON_LOSE_ORGAN, COMSIG_LIVING_REVIVE, COMSIG_LIVING_DEATH, SIGNAL_ADDTRAIT(TRAIT_NOBLOOD), SIGNAL_REMOVETRAIT(TRAIT_NOBLOOD)))
+
+/datum/component/manual_heart/proc/restart()
+	SIGNAL_HANDLER
+
+	if(check_valid())
+		last_pump = world.time
+		pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS) //make sure the action button always shows as available when it is
+		START_PROCESSING(SSdcs, src)
+
+/datum/component/manual_heart/proc/pause()
+	SIGNAL_HANDLER
+	pump_action.build_all_button_icons(UPDATE_BUTTON_STATUS)
+	STOP_PROCESSING(SSdcs, src)
+
+/// Worker proc that checks logic for if a pump can happen, and applies effects/notifications from doing so
+/datum/component/manual_heart/proc/on_pump(mob/owner)
+	last_pump = world.time
+	playsound(owner,'sound/effects/singlebeat.ogg', 40, TRUE)
+	owner.balloon_alert(owner, "your heart beats")
+
+	if(!iscarbon(owner))
+		return
+	var/mob/living/carbon/carbon_owner = owner
+
+	if(HAS_TRAIT(carbon_owner, TRAIT_NOBLOOD))
+		return
+	carbon_owner.blood_volume = min(carbon_owner.blood_volume + (blood_loss * 0.5), BLOOD_VOLUME_MAXIMUM)
+	carbon_owner.remove_client_colour(/datum/client_colour/manual_heart_blood)
+	add_colour = TRUE
+	carbon_owner.adjustBruteLoss(-heal_brute)
+	carbon_owner.adjustFireLoss(-heal_burn)
+	carbon_owner.adjustOxyLoss(-heal_oxy)
+
+/datum/component/manual_heart/process()
+	var/mob/living/carbon/carbon_parent = parent
+
+	//If they aren't connected, don't kill them.
+	if(!istype(carbon_parent) || !carbon_parent.client)
+		last_pump = world.time
+		return
+
+	if(world.time <= (last_pump + pump_delay))
```

Comment:
Could this be replaced with `COOLDOWN_START/COOLDOWN_DECLARE/COOLDOWN_FINISHED`? It looks like these helpers do what you are trying to do.